streaming/src/main/scala/org/apache/spark/streaming/util/BatchedWriteAheadLog.scala

```diff
@@ -0,0 +1,212 @@
+
+/**
+ * A wrapper for a WriteAheadLog that batches records before writing data. All other methods will
+ * be passed on to the wrapped class.
+ *
+ * Parent exposed for testing.
+ */
+private[streaming] class BatchedWriteAheadLog(private[util] val parent: WriteAheadLog)
+  extends WriteAheadLog with Logging {
+
+  import BatchedWriteAheadLog._
+
+  // exposed for tests
+  protected val walWriteQueue = new LinkedBlockingQueue[RecordBuffer]()
+
+  private val WAL_WRITE_STATUS_TIMEOUT = 5000 // 5 seconds
+
+  // Whether the writer thread is active
+  @volatile private var active: Boolean = true
+  protected val buffer = new ArrayBuffer[RecordBuffer]()
+
+  private val batchedWriterThread = startBatchedWriterThread()
+
+  /**
+   * Write a byte buffer to the log file. This method adds the byteBuffer to a queue and blocks
+   * until the record is properly written by the parent.
+   */
+  override def write(byteBuffer: ByteBuffer, time: Long): WriteAheadLogRecordHandle = {
+    val promise = Promise[WriteAheadLogRecordHandle]()
+    walWriteQueue.offer(RecordBuffer(byteBuffer, time, promise))
+    try {
+      Await.result(promise.future.recover { case _ => null }(ThreadUtils.sameThread),
+        WAL_WRITE_STATUS_TIMEOUT.milliseconds)
+    } catch {
+      case e: TimeoutException =>
+        logWarning(s"Write to Write Ahead Log promise timed out after " +
+          s"$WAL_WRITE_STATUS_TIMEOUT millis for record.")
+        null
+    }
+  }
+
+  /**
+   * Read a segment from an existing Write Ahead Log. The data may be aggregated, and the user
+   * should de-aggregate using [[BatchedWriteAheadLog.deaggregate]]
+   *
+   * This method is handled by the parent WriteAheadLog.
+   */
+  override def read(segment: WriteAheadLogRecordHandle): ByteBuffer = {
+    parent.read(segment)
+  }
+
+  /**
+   * Read all the existing logs from the log directory.
+   *
+   * This method is handled by the parent WriteAheadLog.
+   */
+  override def readAll(): JIterator[ByteBuffer] = {
+    parent.readAll().asScala.flatMap(deaggregate).asJava
+  }
+
+  /**
+   * Delete the log files that are older than the threshold time.
+   *
+   * This method is handled by the parent WriteAheadLog.
+   */
+  override def clean(threshTime: Long, waitForCompletion: Boolean): Unit = {
+    parent.clean(threshTime, waitForCompletion)
+  }
+
+
+  /**
+   * Stop the batched writer thread, fulfill promises with failures and close parent writer.
+   */
+  override def close(): Unit = {
+    logInfo("BatchedWriteAheadLog shutting down.")
+    active = false
+    batchedWriterThread.interrupt()
+    batchedWriterThread.join()
+    fulfillPromises()
+    parent.close()
+  }
+
+  /**
+   * Respond to any promises that may have been left in the queue, to unblock receivers during
+   * shutdown.
+   */
+  private def fulfillPromises(): Unit = {
+    while (!walWriteQueue.isEmpty) {
+      val RecordBuffer(_, _, promise) = walWriteQueue.poll()
+      promise.success(null)
```

Comment:
It is weird to signify failure by saying `promise.success(null)`. Rather better to fail the promise by passing on different reasons, and let the promise-waiting-code handle the failure. Otherwise, there are multiple places that are signifying failures using `promise.success`, and then there is a `promise.recover` which is handle failures elsewhere ... not very intuitive.